sycl/include/CL/sycl/atomic_ref.hpp

```diff
@@ -0,0 +1,662 @@
+//==----- atomic_ref.hpp - SYCL 2020 atomic_ref ----------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/__spirv/spirv_ops.hpp>
+#include <CL/sycl/access/access.hpp>
+#include <CL/sycl/atomic.hpp>
+#include <CL/sycl/detail/defines.hpp>
+#include <CL/sycl/detail/spirv.hpp>
+#include <CL/sycl/memory_enums.hpp>
+#ifdef __SYCL_DEVICE_ONLY__
+#include <CL/sycl/multi_ptr.hpp>
+#endif
+
+#ifndef __SYCL_DEVICE_ONLY__
+#include <atomic>
+#endif
+#include <type_traits>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+using memory_order = cl::sycl::memory_order;
+using memory_scope = cl::sycl::memory_scope;
+
+template <typename T> struct IsValidAtomicRefType {
+  static constexpr bool value =
+      (std::is_same<T, int>::value || std::is_same<T, unsigned int>::value ||
+       std::is_same<T, long>::value || std::is_same<T, unsigned long>::value ||
+       std::is_same<T, long long>::value ||
+       std::is_same<T, unsigned long long>::value ||
+       std::is_same<T, float>::value || std::is_same<T, double>::value ||
+       std::is_pointer<T>::value);
+};
+
+// DefaultOrder parameter is limited to read-modify-write orders
+template <memory_order Order>
+using IsValidDefaultOrder = bool_constant<Order == memory_order::relaxed ||
```

Comment:
If I understand correctly, [`bool_constant`](https://en.cppreference.com/w/cpp/types/integral_constant) resides in `std` namespace. The fact that you are able to use it here without `std::` concerns me, because it assumes that there is a `using namespace std;` somewhere before this header is included.

UPD: I've realized that we have `bool_constant` in our `sycl::detail` namespace:
https://github.com/intel/llvm/blob/0c33048e492661e50596eb5f8650eeff92347db9/sycl/include/CL/sycl/detail/stl_type_traits.hpp#L39-L40

That is probably a minor thing, but it still confusing: it is not clear which version is intended to be used here, we don't have explicit include for an internal one, etc.

Reply:
Since the `bool_constant` is used here, in the `sycl::detail` namespace, the definition from the same namespace will be used, so our crafted `bool_constant`. As I understand, we defined `sycl::detail::bool_constant` since this alias in the `std` namespace has been defined since C++17 (https://en.cppreference.com/w/cpp/types/integral_constant) but we also should support C++14 too.